File: android/src/main/java/com/sendbird/calls/reactnative/RNCallsInternalError.kt
package com.sendbird.calls.reactnative

class RNCallsInternalError(from: String? = "unknown", private val type: Type) : Exception(type.message) {
    override val message = "[${from}] ${this.type.message}"
    val code = INTERNAL_ERROR_CODE

    enum class Type(val message: String) {
        // DirectCall
        NOT_FOUND_DIRECT_CALL("There is no DirectCall"),
        NOT_FOUND_VIDEO_DEVICE("Cannot found device with specific id"),
        NOT_FOUND_VIDEO_VIEW("Cannot found video view"),
        INVALID_PARAMS("Invalid parameters"),
        // GroupCall - Room
        NOT_FOUND_ROOM("There is no Room"),
        NOT_FOUND_PARTICIPANT("There is no participant"),
        // others
        NOT_FOUND_QUERY("Cannot found query"),
        UNKNOWN("Unknown error")
    }

    companion object {
        const val INTERNAL_ERROR_CODE = "RNCALLS_INTERNAL"
    }
}

File: ios/Helpers/RNCallsInternalError.swift
//
//  RNCallsInternalError.swift
//  RNSendbirdCalls
//
//  Created by Airen Kang on 2022/06/03.
//  Copyright © 2022 Sendbird. All rights reserved.
//

import Foundation

let INTERNAL_ERROR_CODE = "RNCALLS_INTERNAL"
enum RNCallsInternalError: Error {
    case queryCreateFailure(_ from: String)
    case invalidParams(_ from: String)
    case notFoundVideoDevice(_ from: String)
    case notFoundVideoView(_ from: String)
    case notFoundDirectCall(_ from: String)
    case noResponse(_ from: String)
    case tokenParseFailure(_ from: String)
    case notFoundRoom(_ from: String)
    case unknown(_ from: String)
    
    var message: String {
        switch self {
        case let .queryCreateFailure(from: from):
            return "[\(from)] Create query failure"
        case let .invalidParams(from: from):
            return "[\(from)] Invalid parameters"
        case let .notFoundVideoDevice(from: from):
            return "[\(from)] Cannot found device with specific id]"
        case let .notFoundVideoView(from: from):
            return "[\(from)] Cannot found video view"
        case let .notFoundDirectCall(from: from):
            return "[\(from)] There is no DirectCall"
        case let .noResponse(from: from):
            return "[\(from)] There is no response"
        case let .tokenParseFailure(from: from):
            return "[\(from)] Failed to parse token, check token format"
        case let .notFoundRoom(from: from):
            return "[\(from)] There is no Room"
        case let .unknown(from: from):
            return "[\(from)] Unexpected error"
        }
    }
}
